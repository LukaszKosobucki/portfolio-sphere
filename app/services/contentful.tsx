import { createClient } from "contentful";

export const fetchSpecificEntries = async ({
  content_type,
  title,
  multiple = false,
  order,
}: {
  content_type: string;
  title?: string;
  multiple?: boolean;
  order?: string;
}) => {
  const client = createClient({
    space: process.env.CONTENTFUL_SPACE_ID ?? "",
    environment: "master", // defaults to 'master' if not set
    accessToken: process.env.CONTENTFUL_ACCESS_TOKEN ?? "",
  });

  const query = {
    content_type,
    ...(title && { "fields.title": title }),
    limit: multiple ? 100 : 1,
    order,
  };

  const entries = await client.getEntries(query).then((res) => res.items);

  return entries;
};
