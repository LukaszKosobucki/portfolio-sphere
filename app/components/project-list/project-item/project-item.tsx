import { formatDate } from "@/app/utils/format-date";
import { Entry, EntrySkeletonType } from "contentful";

function ProjectItem({
  entry,
}: {
  entry: Entry<EntrySkeletonType, undefined, string>;
}) {
  return (
    <li>
      <a
        target="_blank"
        href={entry.fields.url as string}
        className="group p-4 gap-3 flex items-center border rounded-lg hover:bg-black/5 hover:dark:bg-white/10 border-black/15 dark:border-white/20 transition-colors duration-300 ease-in-out"
      >
        <div className="w-full group-hover:text-black group-hover:dark:text-white blend">
          <div className="flex flex-wrap items-center gap-2">
            <div className="text-sm uppercase">{formatDate(entry.fields.date as string)}</div>
          </div>
          <div className="font-semibold mt-3 text-black dark:text-white">
            {entry.fields.name as string}
          </div>

          <div className="text-sm line-clamp-2">{entry.fields.description as string}</div>
          <ul className="flex flex-wrap mt-2 gap-1">
            {entry.fields.tags &&
              (entry.fields.tags as Array<string>).map((tag: string) => (
                <li
                  key={tag}
                  className="text-xs uppercase py-0.5 px-1 rounded bg-black/5 dark:bg-white/20 text-black/75 dark:text-white/75"
                >
                  {tag}
                </li>
              ))}
          </ul>
        </div>
        <svg
          xmlns="http://www.w3.org/2000/svg"
          width="20"
          height="20"
          viewBox="0 0 24 24"
          fill="none"
          strokeWidth="2.5"
          strokeLinecap="round"
          strokeLinejoin="round"
          className="stroke-current group-hover:stroke-black group-hover:dark:stroke-white"
        >
          <line
            x1="5"
            y1="12"
            x2="19"
            y2="12"
            className="scale-x-0 group-hover:scale-x-100 translate-x-4 group-hover:translate-x-1 transition-all duration-300 ease-in-out"
          />
          <polyline
            points="12 5 19 12 12 19"
            className="translate-x-0 group-hover:translate-x-1 transition-all duration-300 ease-in-out"
          />
        </svg>
      </a>
    </li>
  );
}

export default ProjectItem;
