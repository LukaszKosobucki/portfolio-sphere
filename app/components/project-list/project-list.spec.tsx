import { fetchSpecificEntries } from "@/app/services/contentful";
import { formatSnapshot } from "@/app/utils/tests/format-snapshot";
import { mockProjectItem } from "@/app/utils/tests/mocks/project-item.mock";
import { render, screen, waitFor } from "@testing-library/react";
import { afterAll, describe, expect, jest, mock, test } from "bun:test";
import ProjectList from "./project-list";

mock.module("@/app/services/contentful", () => ({
  fetchSpecificEntries: jest.fn(),
}));

const mockProjects = [mockProjectItem];
describe("ProjectList Component", () => {
  afterAll(() => {
    jest.clearAllMocks();
  });

  test("should match the snapshot", async () => {
    (fetchSpecificEntries as jest.Mock).mockResolvedValue(mockProjects);

    const { asFragment, getByText } = render(await ProjectList());

    await waitFor(() => {
      expect(getByText("Memory Game")).toBeInTheDocument();
    });

    expect(formatSnapshot(asFragment())).toMatchSnapshot();
  });

  test("should render a list of projects when data is fetched successfully", async () => {
    (fetchSpecificEntries as jest.Mock).mockResolvedValue(mockProjects);

    const { getByText } = render(await ProjectList());

    await waitFor(() => {
      expect(getByText("Memory Game")).toBeInTheDocument();
      expect(
        getByText(
          "Key challenge was mobile optimization. Used tools like Figma, SonarCloud, and Firestore for design, scanning, and leaderboard features.",
        ),
      ).toBeInTheDocument();
    });
    expect(fetchSpecificEntries).toHaveBeenCalledWith({
      content_type: "project",
      multiple: true,
    });
  });

  test("should render correctly when no projects are returned", async () => {
    (fetchSpecificEntries as jest.Mock).mockResolvedValue([]);

    const { getByText } = render(await ProjectList());

    await waitFor(() => {
      const projectItems = screen.queryAllByTestId("projectItem");
      expect(projectItems).toHaveLength(0);

      expect(getByText("Projects")).toBeInTheDocument();
    });
    expect(fetchSpecificEntries).toHaveBeenCalledWith({
      content_type: "project",
      multiple: true,
    });
  });
});
