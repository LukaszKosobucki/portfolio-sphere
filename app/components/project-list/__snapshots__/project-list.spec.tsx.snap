// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`ProjectList Component should match the snapshot 1`] = `
"<DocumentFragment>
  <section
    class="animate scroll-mt-16"
    id="projects"
  >
    <div
      class="space-y-4"
    >
      <div
        class="flex justify-between"
      >
        <p
          class="font-semibold text-black dark:text-white"
        >
          Projects
        </p>
      </div>
      <ul
        class="space-y-4"
      >
        <li>
          <a
            class="group p-4 gap-3 flex items-center border rounded-lg hover:bg-black/5 hover:dark:bg-white/10 border-black/15 dark:border-white/20 transition-colors duration-300 ease-in-out"
            data-testid="projectItem"
            href="https://www.memory-game.lukaszkosobucki.pl/"
            target="_blank"
          >
            <div
              class="w-full group-hover:text-black group-hover:dark:text-white blend"
            >
              <div
                class="flex flex-wrap items-center gap-2"
              >
                <div
                  class="text-sm uppercase"
                >
                  Jan 2023
                </div>
              </div>
              <div
                class="font-semibold mt-3 text-black dark:text-white"
              >
                Memory Game
              </div>
              <div
                class="text-sm line-clamp-2"
              >
                Key challenge was mobile optimization. Used tools like Figma, SonarCloud, and Firestore for design, scanning, and leaderboard features.
              </div>
              <ul
                class="flex flex-wrap mt-2 gap-1"
              >
                <li
                  class="text-xs uppercase py-0.5 px-1 rounded bg-black/5 dark:bg-white/20 text-black/75 dark:text-white/75"
                >
                  Typescript
                </li>
                <li
                  class="text-xs uppercase py-0.5 px-1 rounded bg-black/5 dark:bg-white/20 text-black/75 dark:text-white/75"
                >
                  React
                </li>
                <li
                  class="text-xs uppercase py-0.5 px-1 rounded bg-black/5 dark:bg-white/20 text-black/75 dark:text-white/75"
                >
                  xState
                </li>
                <li
                  class="text-xs uppercase py-0.5 px-1 rounded bg-black/5 dark:bg-white/20 text-black/75 dark:text-white/75"
                >
                  Jest
                </li>
              </ul>
            </div>
            <svg
              class="stroke-current group-hover:stroke-black group-hover:dark:stroke-white"
              fill="none"
              height="20"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2.5"
              viewBox="0 0 24 24"
              width="20"
              xmlns="http://www.w3.org/2000/svg"
            >
              <line
                class="scale-x-0 group-hover:scale-x-100 translate-x-4 group-hover:translate-x-1 transition-all duration-300 ease-in-out"
                x1="5"
                x2="19"
                y1="12"
                y2="12"
              />
              <polyline
                class="translate-x-0 group-hover:translate-x-1 transition-all duration-300 ease-in-out"
                points="12 5 19 12 12 19"
              />
            </svg>
          </a>
        </li>
      </ul>
    </div>
  </section>
</DocumentFragment>"
`;
